### PR TITLE
Support http2 keep-alive

### LINKED ISSUE #68: Allow configuring keepalive and use a reasonable default value
### What happened?

When a connection is idle, and the networking drops (or the server disappears without properly disconnecting), the client happily waits forever on active stream, but no data will ever come again, at least for a long time if TCP keepalive kicks in.

### What did you expect instead?

We should configure the gRPC channel keepalive feature to have some reasonable default, and to be able to override the defaults via the connection URL. We should also check if some default keepalive is already configured in gRPC.

### Affected version(s)

_No response_

### Affected part(s)

Code (part:code)

### Extra information

Coming from:

* https://github.com/frequenz-floss/frequenz-sdk-python/issues/1036
* https://grpc.io/docs/guides/keepalive/

### PR DESCRIPTION
**Pull request by shsms**

### CODE REVIEW

**Review by Marenz — DISMISSED** [review]

File: src/frequenz/client/base/channel.py
Comment (llucax) [inline src/frequenz/client/base/channel.py]: I would use `timedelta` here, as we try to keep high-level types when facing the users (developers).

**Review by llucax — DISMISSED** [review]

File: src/frequenz/client/base/channel.py
Comment (llucax) [reply src/frequenz/client/base/channel.py]: I would mimic how `SslOptions` works. So, adding a separate class for `KeepAliveOptions`, with a `enabled` config option like with to be able to disabled the whole thing and still provide different defaults in case the end user wants to enable it query params, so also allowing overrides via query params, and allow configuring all options instead of leaving some hardcoded (we can provide the current values as defaults).

Comment (shsms) [reply src/frequenz/client/base/channel.py]: I've added `KeepAliveOptions`.  How should the query parameters specify interval?

Comment (llucax) [reply src/frequenz/client/base/channel.py]: Good question. I guess a float with seconds?

**Comment by llucax** [comment]
> Dismissing because the red colour is super unplesant.

:rofl: 

I wish we used it more actually, I try not to use it because I agree if feels a bit aggressive, but it is very useful to indicate some comments really need to be addressed vs. just adding optional comments. In this case I just used it to avoid a merge at least until we discuss the comments as the PR was already approved, but didn't think of just dismissing the approval, that would have been another option :+1:

File: src/frequenz/client/base/channel.py
Comment (llucax) [inline src/frequenz/client/base/channel.py]: I would not allow `None` here, and provide some default, as `enabled` already covers for the disabled case. Otherwise having `enabled=True, interval=None` doesn't make much sense (also the current default is weird, it is `enabled=True` but it is not actually enabled.

For example:
```suggestion
    interval: timedelta = timedelta(seconds=5)
    """The interval between pings."""
```

Comment (shsms) [reply src/frequenz/client/base/channel.py]: There may not be a default keep-alive that works for the general case, but 5s seems excessive.  Nginx waits for 75s, so I'll set it to 60s: http://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout

Comment (shsms) [reply src/frequenz/client/base/channel.py]: > also the current default is weird, it is `enabled=True` but it is not actually enabled.

:D true, I didn't think of that.

File: src/frequenz/client/base/channel.py
Comment (llucax) [inline src/frequenz/client/base/channel.py]: This comment probably also needs update if you don't allow `None` for `interval`.

File: src/frequenz/client/base/channel.py
Comment (llucax) [reply src/frequenz/client/base/channel.py]: Just for reference, Go sets **TCP** keepalive to 15 seconds intervals, [this article](https://www.evanjones.ca/tcp-connection-timeouts.html) does a small analysis about what it means for data transmission:

>The disadvantage is this sends additional keepalive TCP packets for idle connections, but 1 packet every 15 seconds sounds reasonable to me ([the Go team agrees](https://github.com/golang/go/issues/23459); 100k connections = 6.7k packets/sec = 417 kiB/sec = 0.3% of a 1 Gbps Ethernet).

http/2 keepalive probably have quite a bit more overhead, I'm just putting it here for some context and to remember that this is also for detecting broken connections. Probably for could services 1 minute is enough and probably that's most of the services that we'll have, so 60s seems a reasonable default I guess.

As for the "avoid the connection to be dropped because of inactivity" and also for reference, here are some examples of network actors that could drop idle connections:

* [Amazon network load balancers](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/network-load-balancers.html#connection-idle-timeout): After ~6 minutes (350 seconds, configurable). TCP keep-alive is considered network activity, so it can be used to avoid the connection drop.
* [Azure load balancers](https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-tcp-reset#configurable-tcp-idle-timeout): After 4 minutes (configurable from 4 to 100 minutes). TCP keep-alive is also considered network activity, so it can be used to avoid the connection drop.
* Google ignore http/2 keepalive, so for this use case we need to send an application-level packet to keep connections alive.

Comment (shsms) [reply src/frequenz/client/base/channel.py]: done

File: src/frequenz/client/base/channel.py
Comment (shsms) [reply src/frequenz/client/base/channel.py]: done

File: RELEASE_NOTES.md
Comment (llucax) [inline RELEASE_NOTES.md:1]: I would also add a note to `Upgrading` that now keep-alive is enabled by the default options and that the default configuration should probably be tuned according to the client requirements.

File: src/frequenz/client/base/channel.py
Comment (llucax) [inline src/frequenz/client/base/channel.py:149]: I'm still curious why you are not making these configurable, I don't know if it is because you think is a bad idea or because it is not needed now?

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]:
```suggestion
    """Whether HTTP2 keep-alive should be enabled."""
```

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]:
```suggestion
    """The time to wait for a HTTP2 keep-alive response."""
```

**Review by llucax — COMMENTED** [review]
Only a couple of minor things: I think is important to mention keep-alive is enabled by default now in the release notes, but I'm OK with merging without making `max_pings_without_data` and `keepalive_permit_without_calls` for now, just curious about the reasons not to do it.

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]: HTTP2 pings?
```suggestion
    """The interval between HTTP2 pings."""
```

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]: For consistency with `keep_alive` definition
```suggestion
                        defaults.keep_alive.interval
                        if options.keep_alive_interval is None
                        else options.keep_alive_interval
```

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]:
```suggestion
                    defaults.keep_alive.timeout
                    if options.keep_alive_timeout is None
                    else options.keep_alive_timeout
```

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]: To be fair
```suggestion
            timedelta(seconds=float(keep_alive_opts["keep_alive_interval_s"]))
```

File: src/frequenz/client/base/channel.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/base/channel.py]:
```suggestion
            timedelta(seconds=float(keep_alive_opts["keep_alive_timeout_s"]))
```

File: src/frequenz/client/base/channel.py
Comment (shsms) [reply src/frequenz/client/base/channel.py:149]: These are reasonable defaults, and it is unlikely that they will need to change.  I think it would be just too much information to make them configurable now, but we can revisit later if needed.

File: RELEASE_NOTES.md
Comment (shsms) [reply RELEASE_NOTES.md:1]: done.

File: RELEASE_NOTES.md
Comment (daniel-zullo-frequenz) [inline RELEASE_NOTES.md:17]:
```suggestion
- HTTP2 keep-alive is now enabled by default, with an interval of 60 seconds between pings, and a 20 seconds timeout for responses from the service.  These values are configurable and may be updated based on specific requirements.
```

**Review by daniel-zullo-frequenz — APPROVED** [review]

**Comment by daniel-zullo-frequenz** [comment]
I remove it from the queue in case you want to check/fix a typo. And probably you also want to wait for Luca's approval

File: RELEASE_NOTES.md
Comment (shsms) [reply RELEASE_NOTES.md:17]: I think when used as an adjective, it should be singular.  "20 seconds" is a noun, which is not what we want here.

Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md:17]: alright, I didn't know that :angel:

**Pull request merged**
